### ISSUE TITLE
Update just not null fields ?

### ISSUE THREAD

**Issue by justokay** [requery/requery#522]
Hello,
i got two response models of user info, short and full, can i update only not null fields without clearing another ?

**Comment by npurushe** [comment]
If your updating an existing entity and you want to make a field null when it wasn't before, then you have to set it to null. Is that your question? will close for now, but post back if you have more info.

**Comment by davidwhitman** [comment]
I think they are asking the following (which, incidentally, is the same question I want to ask):

If you have an existing entity `User(id=1, firstName=Kim, lastName=)`
and you want to add new data from a different entity that doesn't contain all of the stored data `User(id=1, firstName=, lastName=Sheckles)`
but without destroying the data you already have, is this possible?

You would want to end up with `User(id=1, firstName=Kim, lastName=Sheckles)`.

Basically, is there a way for `update` to merge, rather than replace? Or is the only way to first fetch the existing data, do the merge yourself, and then update?

**Comment by davidwhitman** [comment]
@justokay I accomplished this with some reflection (to avoid needing to update the code if the model changes in the future). I will paste my code, although you'll need to adapt it for your use case and convert to java if needed.

```kotlin
// AppInfoImple and AppInfoEntity both implement the AppInfo interface
    fun put(appInfo: AppInfoImpl): AppInfoEntity {
        val existingItem = db.data.findByKey(AppInfoEntity::class, appInfo.packageName)

        return if (existingItem == null) {
            db.data.insert(mapApiToModel(appInfo))
        } else {
            merge(from = appInfo, to = existingItem)
            db.data.update(existingItem)
        }
    }

    fun put(appInfos: List<AppInfoImpl>) = appInfos.forEach { put(it) }

    /**
     * For each property of [AppInfo], replace the `to` value with the `from` value *iff* the `from` value is not null.
     * Warning: Uses java reflection.
     */
    private fun merge(from: AppInfo, to: AppInfo) {
        val declaredMethods = AppInfo::class.java.declaredMethods
        val getterPrefix = "get"
        val setterPrefix = "set"
        declaredMethods
                .filter { it.name.startsWith(getterPrefix) }
                .filter { it.invoke(from) != null }
                .forEach { getter ->
                    val setter = declaredMethods
                            .first { it.name.startsWith(setterPrefix) && it.name.substring(setterPrefix.count()) == getter.name.substring(getterPrefix.count()) }
                    Timber.v("Setting method ${setter.name} on ${from.title} to non-null value ${getter.invoke(from)}")
                    setter.invoke(to, getter.invoke(from))
                }
    }
```
